nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -101,7 +102,10 @@ def write_loader_conf(profile: str | None, generation: int, specialisation: str
     with open(f"{LOADER_CONF}.tmp", 'w') as f:
         if TIMEOUT != "":
             f.write(f"timeout {TIMEOUT}\n")
-        f.write("default %s\n" % generation_conf_filename(profile, generation, specialisation))
+        if XEN_EFI == "1":
+            f.write("default xen.conf\n")
+        else:
+            f.write("default %s\n" % generation_conf_filename(profile, generation, specialisation))
```

Comment:
Shouldn’t `xen.conf` be based on a generation filename here for proper rollback handling?

Reply:
Well, xen.efi does not support multiple instances of xen.cfg, which means generations are pretty difficult to implement. I sort of assumed rollbacks would work already, as systemd-boot would replace xen.cfg with an earlier file when you rollback, and xen.conf would then point to a xen.cfg that contains an earlier `init=` path.

Reply:
By rollbacks here I was including picking an older generation from the bootloader without actually rolling back the system egneration; sorry to be unspecific.

That limitation is pretty unfortunate; could we work around it by putting `xen.efi` and `xen.cfg` in separate directories by generation?

Reply:
`xen.efi` and `xen.cfg` must be in the same directory as the kernel and the initrd, and that directory must be called whatever the vendor name is.

Reply:
What a pain. I’m guessing it wouldn’t be feasible to patch their UEFI stuff to work with per‐generation file names?

Reply:
Actually, according to the new documentation, it may be possible to pass the -cfg command line option to xen.efi, and there is some special configuration handling based on the filename of the .efi file. I didn't see this documentation before, as I was looking at the wiki, and it said these features don't exist.

The requirement to remain on the same directory as the kernel and initrd remain, but it may be possible to skip the extraEntries stuff and generalise this into a bootspec-compatible generator, just like the kernel system.

Reply:
Unfortunately, that means more python, and my programming knowledge is sorely lacking. Expect terrible code.

Reply:
Update: I haven't forgotten about this, but after further investigation, it did turn out to be far less trivial than I had hoped.

Here's what we need to do:
1. Add a BootSpec extension with the path to `xen.efi` and a `list of string` with the dom0 `bootParams`.
  - There's one BootSpec `@dataclass` per generation. Most generations will point to the same `xen.efi` and have the same `bootParams`.
2. Copy all available `xen.efi`s to `/boot/EFI/nixos`. Their names (`xen-HASH.efi`) should be different than the .cfg files. We could be cheeky and name them `xen-generation-NUMBER.efi`, so they would (hopefully) automatically pick the appropriate `xen-generation-NUMBER.cfg`, allowing us to skip passing the `-cfg` systemd-boot entry option, but that is not an option, as it'd mean we'd need one `xen.efi` per generation, even when they have the same contents binary-wise. The solution is to instead base their hashes on the `$boot` output of the Xen package. ~~(or lobby the UEFI forum to use a filesystem that accepts symlinks)~~
3. Create `xen-generation-NUMBER.cfg`s. There is one per generation, and multiple generations will use the same `xen.efi`. Their syntax is exactly like this:
  ```ini
  [global]
  default=xen

  [xen]
  options=$bootParams
  kernel=$kernelName root=fstab init=$initForThisGeneration $kernelParams
  ramdisk=$initrdName
  ```
4. Create `xen-generation-NUMBER.conf`s: the systemd-boot entries. Their syntax is as follows:
  ```rc
  title title
  sort_key [1]
  efi bootspec.xen
  options -cfg=xen-generation-NUMBER.cfg
  ```
>[!NOTE]
>**[1]:** The ideal solution is to have something like:
>```
>Gen 2 (Xen) (default)
>Gen 2
>Gen 1 (Xen)
>Gen 1
>```
>But I have no idea how to use `sort_key` to do that.
5. Set the default to the latest Xen generation, and we're done!